source/dotnet/Library/AdaptiveCards.Rendering.Wpf/AdaptiveRenderContext.cs

```diff
@@ -286,24 +286,164 @@ public FontColorConfig GetForegroundColors(AdaptiveTextColor textColor)
+            {
+                RowDefinition rowDefinition = new RowDefinition() { Height = GridLength.Auto };
+                if (isVisible)
+                {
+                    rowDefinition = tagContent.NotAutoHeightRowDefinition;
+                }
+
+                // Trying to set the same rowDefinition twice to the same element is not valid, so we have to make a check first
+                if (!(elementIsCurrentlyVisible && isVisible))
+                {
+                    tagContent.ElementContainer.RowDefinitions[tagContent.ViewIndex] = rowDefinition;
+                }
+            }
+
+            // Columns with ColumnDefinition having stars won't hide so we have to set the width to auto
+            if (tagContent.NotAutoWidthColumnDefinition != null)
+            {
+                ColumnDefinition columnDefinition = new ColumnDefinition() { Width = GridLength.Auto };
+                if (isVisible)
+                {
+                    columnDefinition = tagContent.NotAutoWidthColumnDefinition;
+                }
+
+                // Trying to set the same columnDefinition twice to the same element is not valid, so we have to make a check first
+                if (!(elementIsCurrentlyVisible && isVisible))
+                {
+                    tagContent.ElementContainer.ColumnDefinitions[tagContent.ViewIndex] = columnDefinition;
+                }
+            }
+
+        }
+
         public void ToggleVisibility(IEnumerable<AdaptiveTargetElement> targetElements)
         {
+            HashSet<Grid> elementContainers = new HashSet<Grid>();
+
             foreach (AdaptiveTargetElement targetElement in targetElements)
             {
                 var element = LogicalTreeHelper.FindLogicalNode(CardRoot, targetElement.ElementId);
 
                 if (element != null && element is FrameworkElement elementFrameworkElement)
                 {
-                    Visibility visibility = elementFrameworkElement.Visibility;
+                    bool isCurrentlyVisible = (elementFrameworkElement.Visibility == Visibility.Visible);
+
                     // if we read something with the format {"elementId": <id>", "isVisible": true} or we just read the id and the element is not visible
-                    if ((targetElement.IsVisible.HasValue && targetElement.IsVisible.Value) || (!targetElement.IsVisible.HasValue && visibility != Visibility.Visible))
+                    // otherwise if we read something with the format {"elementId": <id>", "isVisible": false} or we just read the id and the element is visible
+                    bool newVisibility = (targetElement.IsVisible.HasValue && targetElement.IsVisible.Value) || (!targetElement.IsVisible.HasValue && !isCurrentlyVisible);
+
+                    TagContent tagContent = GetTagContent(elementFrameworkElement);
+
+                    SetVisibility(elementFrameworkElement, newVisibility, tagContent);
+
+                    if (tagContent != null)
```

Comment:
check just with tagContent for null check

Reply:
C# doesn't allow null checks with only the variable name as C or C++ do

---
In reply to: [280220168](https://github.com/Microsoft/AdaptiveCards/pull/2767#discussion_r280220168) [](ancestors = 280220168)